File: Medium/279. Perfect Squares/PerfectSquares.py
'''
LeetCode Medium: 279. Perfect Squares
https://leetcode.com/problems/perfect-squares/

Check PerfectSquares.cpp for brief solution explanation.
'''
class Solution:
    def numSquares(self, n: int) -> int:
        dp = [0x3f3f3f3f for i in range(n+1)]
        dp[0] = 0
        for i in range(int(sqrt(n))+1):
            for j in range(n+1):
                if j+i*i<=n:
                    dp[j+i*i] = min(dp[j+i*i], dp[j]+1)
        return dp[n]

File: Medium/279. Perfect Squares/PerfectSquares.java
/*
LeetCode Medium: 279. Perfect Squares
https://leetcode.com/problems/perfect-squares/

Check PerfectSquares.cpp for brief solution explanation.
*/
class Solution {
    public int numSquares(int n) {
        int[]dp = new int[n+1];
        Arrays.fill(dp, 0x3f3f3f3f);
        dp[0] = 0;
        for(int i = 1;i*i<=n;i++){
            for(int j = 0;j<=n;j++){
                if(j+i*i<=n){
                    dp[j+i*i] = Math.min(dp[j+i*i], dp[j]+1);
                }
            }
        }
        return dp[n];
    }
}